lib/srv/transport/transportv1/transport.go

```diff
@@ -416,3 +417,78 @@ func (s *sshStream) Send(frame []byte) error {
 
 	return trace.Wrap(s.stream.Send(s.responseFn(frame)))
 }
+
+// ProxyWindowsDesktopSession establishes a connection to a desktop service and proxies raw messages over the stream.
+// The first request from the client must contain a valid dial target
+// before the connection can be established.
+func (s *Service) ProxyWindowsDesktopSession(stream transportv1pb.TransportService_ProxyWindowsDesktopSessionServer) (err error) {
+	ctx := stream.Context()
+
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return trace.BadParameter("failed to find peer")
+	}
+
+	authzContext, err := s.cfg.authzContextFn(p.AuthInfo)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Wait for the first request to arrive with the dial request.
+	req, err := stream.Recv()
+	if err != nil {
+		return trace.Wrap(err, "failed receiving first message")
+	}
+
+	// Validate the target.
+	if req.DialTarget == nil {
+		return trace.BadParameter("first message must contain a dial target")
+	}
+
+	desktopName := req.DialTarget.GetDesktopName()
+	cluster := req.DialTarget.GetCluster()
+
+	clientDst, err := getDestinationAddress(p.Addr, s.cfg.LocalAddr)
+	if err != nil {
+		return trace.Wrap(err, "could get not client destination address; listener address %q, client source address %q", s.cfg.LocalAddr.String(), p.Addr.String())
+	}
+
+	serviceConn, err := s.cfg.Dialer.DialWindowsDesktop(ctx, p.Addr, clientDst, desktopName, cluster, authzContext.Checker)
+	if err != nil {
+		return trace.Wrap(err, "failed to dial target desktop")
+	}
+
+	clientConn, err := streamutils.NewReadWriter(windowsDesktopStream{stream})
```

Comment:
From what I see, `ProxySSH` wraps the client connection in `MonitorConn()`. Do I need it here as well?

Reply:
No, I don't see any connection limits there.